### ISSUE TITLE
Unable to insert element because its parent is not mapped

### ISSUE THREAD

**Issue by algomaster99** [SpoonLabs/diffmin#78]
**PREV**
```java
class Invocation {
    int a = 1;
    void checkForSign() {
        if (a > 0) {
            System.out.println("y");
        } else if (a < 0) {
            System.out.println("z");
        } else {
            System.out.println("x");
        }
    }
}
```

**NEW**
```java
class Invocation {
    int a = 1;
    void checkForSign() {
        if (a > 0) {
            System.out.println("x");
        } else if (a < 0) {
            System.out.println("y");
        } else {
            System.out.println("z");
        }
    }
}
```

The above test case throws the following error which implies that the additional nodes are not inferred by `SpoonMapping`.
```sh
java.lang.IllegalArgumentException: Element not mapped: {
    System.out.println("x");
}
```

**Comment by slarse** [comment]
I'll try to get around to this next week, probably Tuesday.

**Comment by algomaster99** [comment]
Thank you! Meanwhile, I will look into it and probably add more information that might be relevant for us to debug.

**Comment by slarse** [comment]
Yeah, have a look at `SpoonMapping.inferAdditionalMappings`, I think that's where the problem is. Somewhere.

**Comment by algomaster99** [comment]
I tried to debug this for `if` block inside `else-if`. The [`src` and `dst`](https://github.com/SpoonLabs/diffmin/blob/34a4114cab19304a3bb24c3f2637cfc90f54ebd1/src/main/java/com/diffmin/SpoonMapping.java#L88) are as follows:

**Src**
```java
if (a > 0) {
    System.out.println("y");
} else if (a < 0) {
    System.out.println("z");
} else {
    System.out.println("x");
}
```
**Dst**
```java
if (a > 0) {
    System.out.println("x");
} else if (a < 0) {
    System.out.println("y");
} else {
    System.out.println("z");
}
```
When `while` loop in `inferAddiditonalMappings(CtElement, CtElement)` iterates over the children of `src`, it goes over three nodes.
1. `a > 0`
2. `{ System.out.println("y"); }` (`THEN` branch)
3. `ELSE` branch

All of these nodes are already mapped so the loop eventually terminates because [`srcIdx`](https://github.com/SpoonLabs/diffmin/blob/main/src/main/java/com/diffmin/SpoonMapping.java#L99) becomes `3` while the `dstIdx` is still at `0` (`a > 0` of `Dst`). Thus, the `while` loop never even comes across the block `{ System.out.println("x"); }` (in `Dst`) let alone create a mapping of it to some other node.

P.S. 
I commented all the `if-else` condition from the [this block](https://github.com/SpoonLabs/diffmin/blob/34a4114cab19304a3bb24c3f2637cfc90f54ebd1/src/main/java/com/diffmin/SpoonMapping.java#L106-L116) and only kept the following statements. Surprisingly, _all the tests_ passed! GumTree is working really well in matching nodes otherwise had we been more dependent upon `inferAdditionalMappings`, commenting the conditions should have negatively affected the test cases. https://github.com/SpoonLabs/diffmin/blob/34a4114cab19304a3bb24c3f2637cfc90f54ebd1/src/main/java/com/diffmin/SpoonMapping.java#L112-L115

**Comment by slarse** [comment]
> I commented all the if-else condition from the this block and only kept the following statements.

I can guarantee that's not the solution we're looking for. Doing that will cause the method to _remap_ already existing mappings, which is not desirable.

I'm not sure what the solution is here, I'll have to look closer at it next week.

**Comment by algomaster99** [comment]
This is happening in _a lot_ of cases. I can show it to you in our next meeting and then let's hope we come up with a solution. Otherwise, we would have to look for a different way to get the mapping of an element if we can't modify `SpoonMapping.java`.

**Comment by slarse** [comment]
Okay, let's look at it during the next meeting.
